### PR TITLE
ltc: sync from official develop branch

### PR DESCRIPTION
**Pull request by matt2048**
All source files under src directory and header files under include
directory have been synced, but have not added all macros newly added
in official develop branch and the last synced SHA1 in official
develop branch is 6ad52252688bb34f90b5e79da4830a927e87b81f

Signed-off-by: Matt Ma <matt.ma@linaro.org>

### CODE REVIEW

File: core/lib/libtomcrypt/sub.mk [hunk 1]
```diff
@@ -100,6 +100,8 @@ $(eval $(call cryp-dep-one, GCM, AES))
 $(eval $(call cryp-dep-one, AES, ECB CBC CTR CTS XTS))
 # If no DES cipher mode is left, disable DES
 $(eval $(call cryp-dep-one, DES, ECB CBC))
+# If no needed hash algorithm is left, disable DSA
+$(eval $(call cryp-dep-one, DSA, SHA256 SHA384 SHA512))
```
Comment (jforissier) [inline core/lib/libtomcrypt/sub.mk]: Should be `cryp-dep-all`, not `cryp-dep-one`:
```
# dsa_make_params() needs all three SHA-2 algorithms.
# Disable DSA if any is missing.
$(eval $(call cryp-dep-all, DSA, SHA256 SHA384 SHA512))
```

**Comment by jforissier** [comment]
@matt2048 thanks for this big patch ;)
I can't say I have reviewed the LTC code updates in details, but I think you have proceeded as expected. So, this is:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`

**Comment by d3zd3z** [comment]
@matt2048, do you know which was the starting commit the tree was based off of before you brought these in? Since we have also made local modifications to the tree, it might be a good idea for a reviewer to try to verify that this merge hasn't removed any functionality.

I'm a little curious how many upstream commits this brings in.  It looks large, so this might be the best way to do it.  If it was a smaller number, preserving the commits could be useful, but only if it isn't an excessive amount of work.

File: core/lib/libtomcrypt/sub.mk (hunk 1 above)
Comment (matt2048) [inline core/lib/libtomcrypt/sub.mk]: Yes, sorry, should be cryp-dep-all, I will fix it.

**Comment by matt2048** [comment]
@jforissier, yes, I have run xtest on QEMUv8+OP-TEE with the synced LTC, there is no regression for xtest.

**Comment by matt2048** [comment]
@d3zd3z, it is different to find the the starting commit, and we done some synchronization only for some files, such as files under src/pk/asn1/der directory, but not for all source files. I have made a big diff for each source file and header file, and all our own local modifications have been retained.

I think our current code base is very old, such as ff736a61bb39 in official branch as Jerome mentioned in email, was commited two years ago, but we do not have it yet, so before doing the sync job, I have talked with Jerome about how to do the sync job, we have decided to do this big diff at last and this is a comparatively easier way, otherwise, it is difficult to cherry-pick each commit to our code base and there will be too many commits, after all, we focus on it's functionality, not it's history.

**Comment by d3zd3z** [comment]
@matt2048 understandable.

Another question, what do we do about the checkpatch errors? We probably want to leave the tomcrypto code alone, but it does make things looks like it failed in Travis.

**Comment by jforissier** [comment]
@d3zd3z we don't fix the checkpatch errors in imported code, as you can imagine it would just make syncing with upstream more difficult.
I consider the Travis "failed" status a minor annoyance, but I would also welcome a fix if you can find one ;) Such as: excluding some paths maybe? Does checkpatch have such a capability?

**Comment by d3zd3z** [comment]
I don't know of a way to exclude in checkpatch. The only thing I can think of is somehow excluding certain paths from the output of format-patch. I don't think format-patch supports that either, so it would have to be a script the pipe goes through.

It would be nice, though, since it can be a little tedious to figure out what went wrong in travis.

**Comment by jforissier** [comment]
@d3zd3z checkpatch issue hopefully addressed in https://github.com/OP-TEE/optee_os/pull/993.

**Comment by jforissier** [comment]
I have restarted the Travis build to confirm that https://github.com/OP-TEE/optee_os/pull/993 fixes the checkpatch issue. And indeed it's quiet now.

**Comment by jforissier** [comment]
@d3zd3z since you have commented: do you want to add your Acked-by or Reviewed-by? Or should we proceed without?

**Comment by jbech-linaro** [comment]
@matt2048 , thanks for the documentation commit which I asked for, however, that doesn't belong on this particular pull request. Please create a separate pull request for that.

**Comment by jbech-linaro** [comment]
@d3zd3z would you mind adding either an Acked-by or Reviewed-by? See Jerome's comment above. I think this LTC rebase is ready for a quick rebase then merge.

**Comment by matt2048** [comment]
@jbech-linaro , OK, I will do it in a separate pull request, thanks.

**Comment by d3zd3z** [comment]
```
Reviewed-by: David Brown <david.brown@linaro.org>
```

**Comment by matt2048** [comment]
Thanks d3zd3z, I have done the rebase process against the OP-TEE official master branch and added the review tag from  d3zd3z.

**Comment by jforissier** [comment]
Merged, thanks!

**Comment by jbech-linaro** [comment]
Thanks @matt2048  and @jforissier! This is good, since from now on we can probably cherry-pick patches more easily from the official LTC project.

**Pull request merged**